[SPARK-19311][SQL] fix UDT hierarchy issue

gmoehler opened the pull request:
## What changes were proposed in this pull request?
acceptType() in UDT will no only accept the same type but also all base types

## How was this patch tested?
Manual test using a set of generated UDTs fixing acceptType() in my user defined types
*Update: I added a test case to https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala

Please review http://spark.apache.org/contributing.html before opening a pull request.

Commit:
fix UDT hierarchy issue https://issues.apache.org/jira/browse/SPARK-19311

AmplabJenkins commented:
Can one of the admins verify this patch?

viirya commented:
cc @cloud-fan @rxin Although UDT is now private API, some developers still use it by defining their codes as spark package. I am not sure if you think we need to fix this or not.

cloud-fan commented:
is it possible to add a unit test? the change LGTM

viirya commented:
LGTM too. @gmoehler Can you add a unit test?

gatorsmile commented:
ok to test

gatorsmile commented:
You can add a test case in https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala

SparkQA commented:
**[Test build #71755 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71755/testReport)** for PR 16660 at commit [`7ea9aa6`](https://github.com/apache/spark/commit/7ea9aa636f430a30b8d83ed2dda954fd06347d79).

SparkQA commented:
**[Test build #71755 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71755/testReport)** for PR 16660 at commit [`7ea9aa6`](https://github.com/apache/spark/commit/7ea9aa636f430a30b8d83ed2dda954fd06347d79).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71755/
Test PASSed.

Commit:
add test case for SPARK-19311. Test case still fails - cause from elsewhere?

Test case failure is:
- SPARK-19311: UDFs disregard UDT type hierarchy *** FAILED ***
  org.apache.spark.sql.catalyst.errors.package$TreeNodeException: Max iterations (100) reached for batch Resolution, tree:
Project [UDF(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(cast(UDF(41) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType) as exampleBaseType)) AS UDF(UDF(41))#166]
+- SubqueryAlias tmp_table
   +- Project [_1#157 AS id#160, _2#158 AS saying#161]
      +- LocalRelation [_1#157, _2#158]
  at org.apache.spark.sql.catalyst.rules.RuleExecutor$$anonfun$execute$1.apply(RuleExecutor.scala:105)
  at org.apache.spark.sql.catalyst.rules.RuleExecutor$$anonfun$execute$1.apply(RuleExecutor.scala:74)
  at scala.collection.immutable.List.foreach(List.scala:381)
  at org.apache.spark.sql.catalyst.rules.RuleExecutor.execute(RuleExecutor.scala:74)
  at org.apache.spark.sql.execution.QueryExecution.analyzed$lzycompute(QueryExecution.scala:64)
  at org.apache.spark.sql.execution.QueryExecution.analyzed(QueryExecution.scala:62)
  at org.apache.spark.sql.execution.QueryExecution.assertAnalyzed(QueryExecution.scala:48)
  at org.apache.spark.sql.Dataset$.ofRows(Dataset.scala:64)
  at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:592)
  at org.apache.spark.sql.test.SQLTestUtils$$anonfun$sql$1.apply(SQLTestUtils.scala:61)

SparkQA commented:
**[Test build #71841 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71841/testReport)** for PR 16660 at commit [`fad9f0e`](https://github.com/apache/spark/commit/fad9f0e1bc98ad5e7ee2c8dfc726ddc979204f11).

SparkQA commented:
**[Test build #71841 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71841/testReport)** for PR 16660 at commit [`fad9f0e`](https://github.com/apache/spark/commit/fad9f0e1bc98ad5e7ee2c8dfc726ddc979204f11).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  sealed trait IExampleBaseType extends Serializable `
  * `  class ExampleBaseClass(override val field: Int) extends IExampleBaseType `
  * `  class ExampleSubClass(override val field: Int)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71841/
Test FAILed.

Commit:
minor corrections of unit test case

SparkQA commented:
**[Test build #71846 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71846/testReport)** for PR 16660 at commit [`fb261d7`](https://github.com/apache/spark/commit/fb261d7e329b0da5970fd7f6f7eaa2ee5064d89b).

SparkQA commented:
**[Test build #71846 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71846/testReport)** for PR 16660 at commit [`fb261d7`](https://github.com/apache/spark/commit/fb261d7e329b0da5970fd7f6f7eaa2ee5064d89b).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71846/
Test FAILed.

viirya commented on sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala:
```diff
@@ -71,6 +71,94 @@ object UDT {
 
+// object and classes to test SPARK-19311
+
+  // Trait/Interface for base type
```
Please fix the indentation.

gmoehler replied:
corrected in latest commit ... thanks for pointing out

Commit:
re-organize imports to satisfy Scalastyle checks

SparkQA commented:
**[Test build #71849 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71849/testReport)** for PR 16660 at commit [`6b6b773`](https://github.com/apache/spark/commit/6b6b773a4df1ee017f00b9ca43a76a429ad418de).

SparkQA commented:
**[Test build #71849 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71849/testReport)** for PR 16660 at commit [`6b6b773`](https://github.com/apache/spark/commit/6b6b773a4df1ee017f00b9ca43a76a429ad418de).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71849/
Test FAILed.

Commit:
re-organize imports to satisfy Scalastyle checks & add correct indendation & lf

SparkQA commented:
**[Test build #71852 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71852/testReport)** for PR 16660 at commit [`4ae0555`](https://github.com/apache/spark/commit/4ae05559d9a1b343e61e87ad0e4841c497c13382).

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala:
```diff
@@ -71,7 +73,104 @@ object UDT {
+
+// a base class
+class ExampleBaseClass(override val field: Int) extends IExampleBaseType {
+  override def toString: String = field.toString
```
why override `toString` here?

gatorsmile replied:
Please simplify it to 
```
class ExampleBaseClass(override val field: Int) extends IExampleBaseType
```

viirya replied:
@gmoehler I think we don't need `toString`?

gmoehler replied:
ok

SparkQA commented:
**[Test build #71852 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71852/testReport)** for PR 16660 at commit [`4ae0555`](https://github.com/apache/spark/commit/4ae05559d9a1b343e61e87ad0e4841c497c13382).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `sealed trait IExampleBaseType extends Serializable `
  * `class ExampleBaseClass(override val field: Int) extends IExampleBaseType `
  * `class ExampleSubClass(override val field: Int)`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71852/
Test PASSed.

gatorsmile commented on sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala:
```diff
@@ -20,13 +20,15 @@ package org.apache.spark.sql
 import org.apache.spark.sql.test.SharedSQLContext
 import org.apache.spark.sql.types._
 
+
```
Nit: Please remove this empty line for avoiding unnecessary changes.

gmoehler replied:
ok

gatorsmile commented on sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala:
```diff
@@ -71,7 +73,104 @@ object UDT {
+  extends ExampleBaseClass(field) with IExampleSubType
+
+// UDT for base class
+private[spark] class ExampleBaseTypeUDT extends UserDefinedType[IExampleBaseType] {
```
`private[spark] ` is not needed here

gmoehler replied:
ok

gatorsmile commented on sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala:
```diff
@@ -71,7 +73,104 @@ object UDT {
+
+  override def typeName: String = "exampleBaseType"
+
+  private[spark] override def asNullable: ExampleBaseTypeUDT = this
```
Nit: to simplify the test cases, please remove `asNullable `, `typeName `, `equals` and `hashCode`

gmoehler replied:
ok

gatorsmile commented on sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala:
```diff
@@ -71,7 +73,104 @@ object UDT {
+  }
+
+  override def serialize(obj: IExampleSubType): InternalRow = {
+
```
Nit: Please remove this line. Thanks!

gmoehler replied:
not sure which line you mean... we need to overwrite serialize in any case I guess - or is it about a blank line?

gatorsmile commented on sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala:
```diff
@@ -71,7 +73,104 @@ object UDT {
+    datum match {
+      case row: InternalRow =>
+        require(row.numFields == 1,
+          s"VectorUDT.deserialize given row with length " +
```
Please remove this unneeded `s` String Interpolator. Thanks!

gmoehler replied:
ok - replaced it with a fixed string

gatorsmile commented on sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala:
```diff
@@ -71,7 +73,104 @@ object UDT {
+
+  override def typeName: String = "exampleSubType"
+
+  private[spark] override def asNullable: ExampleSubTypeUDT = this
```
Nit: to simplify the test cases, please remove asNullable, typeName, equals and hashCode too.

gmoehler replied:
done

gatorsmile commented on sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala:
```diff
@@ -71,7 +73,104 @@ object UDT {
+
+
 class UserDefinedTypeSuite extends QueryTest with SharedSQLContext with ParquetTest {
+
```
Nit: Please remove this empty line.

gmoehler replied:
ok

gatorsmile commented:
It looks great! My comment is just to simplify the unit test cases. Normally, we want to make the unit test cases as simple as possible. Thanks!

LGTM except a few minor comments.

viirya commented on sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala:
```diff
@@ -194,4 +293,35 @@ class UserDefinedTypeSuite extends QueryTest with SharedSQLContext with ParquetT
   }
+
+  test("SPARK-19311: UDFs disregard UDT type hierarchy") {
+    UDTRegistration.register(classOf[IExampleBaseType].getName,
```
With `SQLUserDefinedType`, no need to use `UDTRegistration`. We can remove this two lines.

gmoehler replied:
i tend to leave them, but remove the @SQLUserDefinedType, so we have a test that uses UDTRegistration

viirya replied:
oh. if you worry about that, actually we have `UDTRegistrationSuite` for test case of `UDTRegistration`. i am fine to either `SQLUserDefinedType` or `UDTRegistration`.

gmoehler replied:
ok

viirya commented on sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala:
```diff
@@ -71,7 +73,104 @@ object UDT {
+
+  override def hashCode(): Int = classOf[ExampleSubTypeUDT].getName.hashCode()
+
+  override def equals(other: Any): Boolean = other.isInstanceOf[ExampleSubTypeUDT]
```
`equals` in `UserDefinedType` has default impl. to call `acceptsType`.

gmoehler replied:
right - i have removed equals anyway due to a prev comment of @gatorsmile.

Commit:
worked in comments regarding style & not required overrides

gmoehler commented:
Thanks for the valuable (and fast!) comments - i have worked them in.

SparkQA commented:
**[Test build #71929 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71929/testReport)** for PR 16660 at commit [`7aed9a4`](https://github.com/apache/spark/commit/7aed9a4fada263785ce1d81acb31073ef7a401fd).

SparkQA commented:
**[Test build #71929 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71929/testReport)** for PR 16660 at commit [`7aed9a4`](https://github.com/apache/spark/commit/7aed9a4fada263785ce1d81acb31073ef7a401fd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ExampleBaseTypeUDT extends UserDefinedType[IExampleBaseType] `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71929/
Test PASSed.

viirya commented:
LGTM except one minor comment.

gmoehler commented:
@viirya Which comment are you referring to? I thought i had included all of them ;-)

gatorsmile commented:
The remaining comment is: https://github.com/apache/spark/pull/16660#discussion_r97427591

gatorsmile commented:
LGTM except one comment. Thanks for working on this!

viirya commented:
@gmoehler `ExampleBaseClass` overrides `toString` impl. That is redundant.

Commit:
remove unnecessary toString() method

gmoehler commented:
Thanks for pointing out. I had overseen this comment. 
Really appreciate you commitment to clean code!
Change is pushed and ready for a test build.

SparkQA commented:
**[Test build #71980 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71980/testReport)** for PR 16660 at commit [`6c16760`](https://github.com/apache/spark/commit/6c16760f0b2e5aba864366c03888666ba584d38a).

SparkQA commented:
**[Test build #71980 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71980/testReport)** for PR 16660 at commit [`6c16760`](https://github.com/apache/spark/commit/6c16760f0b2e5aba864366c03888666ba584d38a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ExampleBaseClass(override val field: Int) extends IExampleBaseType`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71980/
Test PASSed.

gatorsmile commented:
LGTM

gatorsmile commented:
Thanks! Merging to master/2.1
